src/fal/planner/plan.py

```diff
@@ -0,0 +1,75 @@
+from __future__ import annotations
+
+from typing import Iterator, List
+
+import networkx as nx
+from fal.node_graph import NodeKind
+
+
+def _find_subgraphs(graph: nx.DiGraph) -> Iterator[List[str]]:
+    # Initially topologically sort the graph and find nodes
+    # that can be grouped together to be run as a single node
+    # by using the critical node approach. All nodes within a
+    # group must have the same ancestors, to avoid removing
+    # existing branches.
+
+    current_stack = []
+    allowed_ancestors = set()
+
+    def split() -> Iterator[List[str]]:
+        if len(current_stack) > 1:
+            yield current_stack.copy()
+        current_stack.clear()
+        allowed_ancestors.clear()
+
+    for node in nx.topological_sort(graph):
+        properties = graph.nodes[node]
+        if properties["kind"] is NodeKind.FAL_MODEL:
+            yield from split()
+            continue
+
+        ancestors = nx.ancestors(graph, node)
+        if not current_stack:
+            allowed_ancestors = ancestors
+
+        if not ancestors.issubset(allowed_ancestors):
+            yield from split()
```

Comment:
Why don't we continue here?

Reply:
We don't need it, if the ancestor check fails we are detaching from the current stack and starting a new one.

Reply:
We do use `continue` in the Python models because it is theoretically impossible for them to be the start of a new stack (since they would cut off immediately).